### PR TITLE
torch/config: fix mock behaviour

### PR DESCRIPTION
**Pull request by c00w**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #140779

Mock only replaces the value that was removed, if after deletion, it
does not see the attribute.

### CODE REVIEW

**Comment by ezyang** [comment]
Naw man, I don't understand. Doesn't this mean that mock will irrecoverably trash the config (hide = True) when it exits scope?

**Comment by c00w** [comment]
@ezyang  - The code to blame is https://github.com/python/cpython/blob/94a7a4e22fb8f567090514785c69e65298acca42/Lib/unittest/mock.py#L1637

TL;DR - Technically we'll have an intermediate state of hiding the config while mock.patch is unpatching itself, but it calls setattr after the delete call.

Essentially, mock.patch requires, that if __dict__ isn't accesible (which it isn't), that after delattr is called on the object, the object must throw when hasattr is called. Otherwise, it doesn't call setattr again.

I've filed a upstream bug at https://github.com/python/cpython/issues/126886 to see if there is any appetite for expanding / fixing this. But even then, we'll have to wait until the fix is in all versions of pytorch.

If this isn't super clear, I could make trying to patch this code throw (specifically by raising an exception if delattr is called), and rewrite all instances with this pattern into using config.patch instead of unittest.mock.patch.

Let me know what you think.

**Comment by kiukchung** [comment]
I can confirm that this fix does indeed make `dynamo/test_exc.py` pass when the entire test file is run as a single `pytest` where the test cases are run in function name alphanumeric order. Without this change `dynamo/test_exc.py` would fail the test case `test_not_implemented_error` because `torch._inductor.config.suppress_errors` would be set to `False` by the one of the previous test cases.

**Review by ezyang — APPROVED** [review]
OK, just put that in the comment. And good thing we're not letting people use del lol
